cmd/gateway-gcs.go

```diff
@@ -312,8 +313,27 @@ func (l *gcsGateway) ListObjects(bucket string, prefix string, marker string, de
 	nextMarker := ""
 	prefixes := []string{}
 
-	// we'll set marker to continue
-	it.PageInfo().Token = marker
+	// To accommodate S3-compatible applications using
+	// ListObjectsV1 to use object keys as markers to control the
+	// listing of objects, we use the following encoding scheme to
+	// distinguish between GCS continuation tokens and application
+	// supplied markers.
+	//
+	// - NextMarker in ListObjectsV1 response is constructed by
+	//   prefixing "##minio" to the GCS continuation token,
+	//   e.g, "##minioCgRvYmoz"
+	//
+	// - Application supplied markers are used as-is to list
+	//   object keys that appear after it in the lexicographical order.
+
+	userMarker := false
+	// If application is using GCS continuation token we should
+	// strip the minioTokenPrefix we added.
+	if strings.HasPrefix(marker, minioTokenPrefix) {
```

Comment:
Can this be converted into a GCS specific function with varying inputs tested? for hasPrefix.